File: worker_py/llm/ollama.py
"""Ollama adapter — fully local inference. Port of ``llm/ollama.ts``."""

import os
from typing import Optional

from ..http import request_with_retry
from .base import LLMProvider


class OllamaLLM(LLMProvider):
    name = "ollama"

    def __init__(self, model: Optional[str] = None, host: Optional[str] = None) -> None:
        self.model = model or os.environ.get("OLLAMA_MODEL") or "llama3.1"
        self.host = (host or os.environ.get("OLLAMA_HOST") or "http://localhost:11434").rstrip("/")

    async def complete(
        self,
        prompt,
        *,
        system=None,
        max_tokens=1024,
        temperature=0.3,
        json=False,
        model=None,
        timeout_ms=120000,
    ) -> str:
        messages = []
        if system:
            messages.append({"role": "system", "content": system})
        messages.append({"role": "user", "content": prompt})

        body = {
            "model": model or self.model,
            "stream": False,
            "options": {"temperature": temperature, "num_predict": max_tokens},
            "messages": messages,
        }
        if json:
            body["format"] = "json"

        res = await request_with_retry(
            "POST",
            f"{self.host}/api/chat",
            headers={"content-type": "application/json"},
            json=body,
            timeout_ms=timeout_ms,
            attempts=2,
        )
        data = res.json()
        try:
            return str((data.get("message") or {}).get("content") or "").strip()
        except AttributeError:
            return ""

File: src/lib/core/llm/ollama.ts
import { fetchWithRetry } from "../http";
import type { LLMCompleteOptions, LLMProvider } from "./types";

/**
 * Ollama adapter — fully local, free, zero-API-key inference. Run `ollama serve`
 * and pull a model (e.g. `ollama pull llama3.1`). This is the recommended default
 * for self-hosting / open-source users who don't want to pay for a hosted API.
 */
export class OllamaLLM implements LLMProvider {
  readonly name = "ollama";
  readonly model: string;
  private readonly host: string;

  constructor(model?: string, host?: string) {
    this.model = model || process.env.OLLAMA_MODEL || "llama3.1";
    this.host = (host || process.env.OLLAMA_HOST || "http://localhost:11434").replace(/\/$/, "");
  }

  async complete(prompt: string, opts: LLMCompleteOptions = {}): Promise<string> {
    const res = await fetchWithRetry(
      `${this.host}/api/chat`,
      {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({
          model: opts.model || this.model,
          stream: false,
          ...(opts.json ? { format: "json" } : {}),
          options: { temperature: opts.temperature ?? 0.3, num_predict: opts.maxTokens ?? 1024 },
          messages: [
            ...(opts.system ? [{ role: "system", content: opts.system }] : []),
            { role: "user", content: prompt },
          ],
        }),
      },
      { timeoutMs: opts.timeoutMs ?? 120000, attempts: 2 }
    );

    const json = await res.json();
    return String(json?.message?.content ?? "").trim();
  }
}
